### DISCUSSION TITLE
GPU jpeg/jpeg2000

### DISCUSSION THREAD

**Discussion by boskar** [libvips/libvips#1993] [Ideas]
Are You planning to add GPU support?

I've found an [interesting jpeg2000 GPU library](http://apps.man.poznan.pl/trac/jpeg2k/wiki) which could work well in some cases. 
There is [another one](http://cuj2k.sourceforge.net/).
I also get accross several jpeg GPU libraries.
http://sourceforge.net/projects/cujpeg/
http://sourceforge.net/p/gpujpeg/code/ci/master/tree/ 

Is it even possible to integrate VIPS with one of them?

**Comment by jcupitt** [comment]
Hi boskar, libvips uses the IJG libjpeg library. You can also link against libjpegturbo, which uses SSE for a useful speedup. These other jpeg libraries have a different API and you'd need to make a new import function for libvips to use one of them.

For jp2k, libvips uses openslide and openslide in turn uses openjpeg. You could try asking the openslide people if they have any plans for GPU acceleration. I know the openjpeg people have been looking at this. 

libvips has no GPU code at the moment. It would be an interesting thing to experiment with: libvips's architecture, with fairly small tiles flowing through a long processing pipeline, ought to be a good fit for GPUs. We could leave this issue open as a reminder to look into this. 

John

**Comment by boskar** [comment]
Good to hear that it could work, somewhere in the future. I've also found a commercial cuda jpeg library, maybe this version is API compatibile. 

Jp2k input is unnecessery for me - I rather thought about jp2k output from openslide input, as intermediate file for iipserver (i'm just using pyramid tiff at the moment). 

I've seen some OpenCL work in libjpeg-turbo, maybe that's the right way.
